src/Raven.Server/Documents/CountersStorage.cs

```diff
@@ -163,162 +173,598 @@ public long GetNumberOfCountersToProcess(DocumentsOperationContext context, stri
+                var newEntry = (CounterValues*)newVal.Ptr + dbIdIndex;
 
-            return (doc, name);
+                newEntry->Value = value;
+                newEntry->Etag = newETag;
+
+                counters.Modifications = new DynamicJsonValue(counters)
+                {
+                    [name] = new BlittableJsonReaderObject.RawBlob
+                    {
+                        Length = newVal.Length,
+                        Ptr = newVal.Ptr
+                    }
+                };
+            }
+
+            if (overrideExisting == false)
+                _countersCount++;
         }
 
-        private static ReplicationBatchItem CreateReplicationBatchItem(DocumentsOperationContext context, Table.TableValueHolder result)
+        private static void IncrementExistingCounter(DocumentsOperationContext context, string documentId, string name, long delta, ref bool exists, BlittableJsonReaderObject.RawBlob existingCounter,
+            int dbIdIndex, long newETag, BlittableJsonReaderObject counters, long value)
         {
-            var (doc, name) = ExtractDocIdAndName(context, result.Reader);
+            var existingCount = existingCounter.Length / SizeOfCounterValues;
 
-            return new ReplicationBatchItem
+            if (dbIdIndex < existingCount)
             {
-                Type = ReplicationItemType.Counter,
-                Id = doc,
-                Etag = TableValueToEtag((int)CountersTable.Etag, ref result.Reader),
-                Name = name,
-                ChangeVector = TableValueToString(context, (int)CountersTable.ChangeVector, ref result.Reader),
-                Value = TableValueToLong((int)CountersTable.Value, ref result.Reader),
-                Collection = TableValueToId(context, (int)CountersTable.Collection, ref result.Reader),
-                TransactionMarker = TableValueToShort((int)CountersTable.TransactionMarker, nameof(ReplicationBatchItem.TransactionMarker), ref result.Reader),
-            };
-        }
+                exists = true;
+                var counter = (CounterValues*)existingCounter.Ptr + dbIdIndex;
+                try
+                {
+                    counter->Value = checked(counter->Value + delta); //inc
+                    counter->Etag = newETag;
+                }
+                catch (OverflowException e)
+                {
+                    CounterOverflowException.ThrowFor(documentId, name, counter->Value, delta, e);
+                }
+            }
+            else
+            {
+                // counter exists , but not with local DbId
 
-        public void PutCounter(DocumentsOperationContext context, string documentId, string collection, string name, long value)
-        {
-            PutCounterImpl(context, documentId, collection, name, null, value);
+                using (AddPartialValueToExistingCounter(context, existingCounter, dbIdIndex, value, newETag))
```

Comment:
This isn't valid, we are disposing the value while still using `existingCounter`

Reply:
fixed